### PR TITLE
IGNITE-9978: [ML] Implement Compound Naive Bayes classifier

### PR DESCRIPTION
**Pull request by dehasi**
Compound Naive Bayes over Gaussian and Discrete models.

To implement Compound Naive Bayes,  Gaussian and Discrete models were updated:
* A method `setFeatureIdsToSkip(Collection<Integer> featureIdsToSkip)`  was added to Gaussian and Discrete trainers. The method accepts a collection with feature ids which should be skipped in a trainer. 

* A method `double[] probabilityPowers(Vector vector)` was added to Gaussian and Discrete models.  The method returns an array where the index is a label index and value is a probability power of being this label.

@zaleslaw @ybabak @avplatonov @dmitrievanthony could you take a look?

### CODE REVIEW

**Comment by zaleslaw** [comment]
I will have a look this week, currently it has conflicts with master

File: examples/src/main/java/org/apache/ignite/examples/ml/naivebayes/CompoundNaiveBayesExample.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
+package org.apache.ignite.examples.ml.naivebayes;
```
Comment (dspavlov) [inline examples/src/main/java/org/apache/ignite/examples/ml/naivebayes/CompoundNaiveBayesExample.java]: Could you please add Apache header containing the license?

File: modules/ml/src/test/java/org/apache/ignite/ml/naivebayes/compound/CompoundNaiveBayesTest.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
+package org.apache.ignite.ml.naivebayes.compound;
```
Comment (dspavlov) [inline modules/ml/src/test/java/org/apache/ignite/ml/naivebayes/compound/CompoundNaiveBayesTest.java]: Could you please add Apache header containing the license?

File: modules/ml/src/test/java/org/apache/ignite/ml/naivebayes/compound/CompoundNaiveBayesTrainerTest.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
+package org.apache.ignite.ml.naivebayes.compound;
```
Comment (dspavlov) [inline modules/ml/src/test/java/org/apache/ignite/ml/naivebayes/compound/CompoundNaiveBayesTrainerTest.java]: Could you please add Apache header containing the license?

File: modules/ml/src/test/java/org/apache/ignite/ml/naivebayes/compound/Data.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
+package org.apache.ignite.ml.naivebayes.compound;
```
Comment (dspavlov) [inline modules/ml/src/test/java/org/apache/ignite/ml/naivebayes/compound/Data.java]: Could you please add Apache header containing the license?

File: examples/src/main/java/org/apache/ignite/examples/ml/naivebayes/CompoundNaiveBayesExample.java (hunk 1 above)
Comment (dehasi) [reply examples/src/main/java/org/apache/ignite/examples/ml/naivebayes/CompoundNaiveBayesExample.java]: I added Apache header in all Compound Bayes tests. Sorry for this oversight

File: modules/ml/src/test/java/org/apache/ignite/ml/naivebayes/compound/CompoundNaiveBayesTrainerTest.java (hunk 3 above)
Comment (dehasi) [reply modules/ml/src/test/java/org/apache/ignite/ml/naivebayes/compound/CompoundNaiveBayesTrainerTest.java]: I added the header.

File: modules/ml/src/test/java/org/apache/ignite/ml/naivebayes/compound/Data.java (hunk 4 above)
Comment (dehasi) [reply modules/ml/src/test/java/org/apache/ignite/ml/naivebayes/compound/Data.java]: ditto.

File: modules/ml/src/test/java/org/apache/ignite/ml/naivebayes/compound/CompoundNaiveBayesTest.java (hunk 2 above)
Comment (dehasi) [reply modules/ml/src/test/java/org/apache/ignite/ml/naivebayes/compound/CompoundNaiveBayesTest.java]: ditto.

**Comment by dehasi** [comment]
@dspavlov @zaleslaw  Thank you for the reviewing. 

I resolved conflicts, it was a formatting problem due to newline separators. I also formatted my code for the same style.

File: modules/ml/src/main/java/org/apache/ignite/ml/naivebayes/discrete/DiscreteNaiveBayesTrainer.java [hunk 5]
```diff
@@ -49,6 +52,9 @@
     /** The threshold to convert a feature to a discrete value. */
     private double[][] bucketThresholds;
 
+    /** Feature ids which should be skipped. By defaut all features are processed. */
+    private Collection<Integer> featureIdsToSkip = emptyList();
+
     /** {@inheritDoc} */
     @Override public <K, V> DiscreteNaiveBayesModel fit(DatasetBuilder<K, V> datasetBuilder,
                                                         Preprocessor<K, V> extractor) {
```
Comment (avplatonov) [inline modules/ml/src/main/java/org/apache/ignite/ml/naivebayes/discrete/DiscreteNaiveBayesTrainer.java]: @dehasi , How do you think, maybe we may remove featureIdsToSkip logic from trainers and models and in CompoundBayess we may create to Preprocessors that wrap input preprocessor by features filtering logic? In this case we avoid modification in Discrete and Gauss models and save this logic in one place.

Comment (dehasi) [reply modules/ml/src/main/java/org/apache/ignite/ml/naivebayes/discrete/DiscreteNaiveBayesTrainer.java]: @avplatonov That's a good idea. I played with `map` function in an extractor,   I can do smth like `extractor = extractor.map(skipFeatures(featureIdsToSkip));` before call `fit`. 
But in this case, I make a copy of each vector.  Is it an acceptable approach?

Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/naivebayes/discrete/DiscreteNaiveBayesTrainer.java]: I think it won't be a big problem. In other case we can use proxy-object like vector-projection that wraps original vector and gives access only to non-filtered features

Comment (dehasi) [reply modules/ml/src/main/java/org/apache/ignite/ml/naivebayes/discrete/DiscreteNaiveBayesTrainer.java]: ok, I used `skipFeatures(featureIdsToSkip)` and reverted Gaussian and Discrete Bayesses.

**Comment by dehasi** [comment]
@avplatonov Thank you for the feedback. 
I moved feature filtering to Compound Bayes.  Not sure if my approach is ok or you want a separate preprocessor? 

Regarding vector-projection, I like the idea itself and if we need this structure I'll be happy to implement it.

If I need to enrich other Bayesian models with `probabilityPowers` method, does it make sense to create a separate interface for it? Something like `CompoundBayesCompatibleModel` and force people to implement this method in (probably) some other types of Bayesian algorithms.

**Comment by dehasi** [comment]
Hi, @avplatonov ,

Thank you for the review.

1. I created a ticket [IGNITE-12105](https://issues.apache.org/jira/browse/IGNITE-12105) to implement vector projecting.

2. I extracted an interface, but I have a question. What do you think, shall the `BayesModel` extend `IgniteModel<>`?

**Comment by avplatonov** [comment]
@dehasi Thanks a lot!
1) Great, thank you.
2) It depends on whether we will add some Bayesian model, which will not be a model in the usual sense. Moreover, IgniteModel is a model that can be serialized and deserialized into well-known and controlled internal classes of Apache Ignite ML. So I think it makes sense to make this an child of IgniteModel.

**Comment by dehasi** [comment]
@avplatonov ,

I made `BayesModel` child of `IgniteModel`.

**Pull request merged**